wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -249,48 +264,45 @@ public Runtime get(String workspaceId) throws NotFoundException, ServerException
     public void stop(String workspaceId, Map<String, String> options) throws NotFoundException,
                                                                              InfrastructureException,
                                                                              ConflictException {
-
+        RuntimeState state = runtimes.get(workspaceId);
+        if (state == null) {
+            throw new NotFoundException("Workspace with id '" + workspaceId + "' is not running.");
+        }
+        if (!state.status.equals(RUNNING)) {
+            throw new ConflictException(
+                    format("Could not start workspace '%s' because its state is '%s'", workspaceId, state.status));
+        }
+        runtimes.replace(workspaceId, new RuntimeState(state.runtime, STOPPING));
```

Comment:
I'm not sure i understand the whole flow, but it looks like this is a place where only the one thread should go through.
I think that this version of replace allows you to achieve such behaviour:
```java
if (!runtimes.replace(workspaceId, state, new RuntimeState(state.runtime, STOPPING))) {
    // some other thread already put the new value, handle it somehow
}
```
WDYT?